File: src/main/java/org/mastodon/mamut/SciviewBridgeUI.java
//TODO add license text
package org.mastodon.mamut;

import org.mastodon.mamut.util.AdjustableBoundsRangeSlider;
import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;
import java.util.ArrayList;
import java.util.function.Consumer;

public class SciviewBridgeUI {
	SciviewBridge controlledBridge;
	final Container contentPane;

	public SciviewBridgeUI(final SciviewBridge controlledBridge, final Container populateThisContainer) {
		this.controlledBridge = controlledBridge;
		this.contentPane = populateThisContainer;
		populatePane();
		updatePaneValues();
	}

	public Container getControlsWindowPanel() {
		return contentPane;
	}

	public SciviewBridge getControlledBridge() {
		return controlledBridge;
	}

	// -------------------------------------------------------------------------------------------
	void populatePane() {
		final GridBagLayout gridBagLayout = new GridBagLayout();
		contentPane.setLayout( gridBagLayout );

		final GridBagConstraints c = new GridBagConstraints();
		c.anchor = GridBagConstraints.LINE_START;
		c.fill = GridBagConstraints.HORIZONTAL;

		c.gridy = 0;
		c.gridwidth = 2;
		c.weightx = 0.1;
		c.insets = new Insets(sideSpace,sideSpace,2,sideSpace);
		contentPane.add( new JLabel("Volume pixel values 'v' are processed linearly, normalized, gamma, scaled back:"), c);
		c.insets = new Insets(2,sideSpace,2,sideSpace);
		c.gridwidth = 1;

		c.gridy++;
		insertNote("   pow( min(contrast*v +shift, not_above)/not_above, gamma ) *not_above", c);

		c.gridy++;
		c.gridx = 0;
		insertLabel("Apply on Volume this contrast factor:", c);
		//
		c.gridx = 1;
		INTENSITY_CONTRAST = new SpinnerNumberModel(1.0, -100.0, 100.0, 0.5);
		insertSpinner(INTENSITY_CONTRAST, (f) -> controlledBridge.INTENSITY_CONTRAST = f, c);

		c.gridy++;
		c.gridx = 0;
		insertLabel("Apply on Volume this shifting bias:", c);
		//
		c.gridx = 1;
		INTENSITY_SHIFT = new SpinnerNumberModel(0.0, -65535.0, 65535.0, 50);
		insertSpinner(INTENSITY_SHIFT, (f) -> controlledBridge.INTENSITY_SHIFT = f, c);

		c.gridy++;
		c.gridx = 0;
		insertLabel("Apply on Volume this gamma level:", c);
		//
		c.gridx = 1;
		INTENSITY_GAMMA = new SpinnerNumberModel(1.0, 0.1, 3.0, 0.1);
		insertSpinner(INTENSITY_GAMMA, (f) -> controlledBridge.INTENSITY_GAMMA = f, c);

		c.gridy++;
		c.gridx = 0;
		insertLabel("Set to zero any voxel whose value would have been above:", c);
		//
		c.gridx = 1;
		INTENSITY_CLAMP_AT_TOP = new SpinnerNumberModel(700.0, 0.0, 65535.0, 50.0);
		insertSpinner(INTENSITY_CLAMP_AT_TOP, (f) -> controlledBridge.INTENSITY_CLAMP_AT_TOP = f, c);

		// -------------- separator --------------
		c.gridy++;
		insertSeparator(c);

		c.gridy++;
		insertNote("Shortcuts to the standard sciview view controls, incl. Volume intensity mapping", c);

		c.gridy++;
		JPanel sliderBarPlaceHolder = new JPanel();
		c.gridx = 0;
		c.gridwidth = 2;
		contentPane.add(sliderBarPlaceHolder, c);
		//
		INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM = AdjustableBoundsRangeSlider.createAndPlaceHere(
				sliderBarPlaceHolder,
				(int)controlledBridge.INTENSITY_RANGE_MIN, (int)controlledBridge.INTENSITY_RANGE_MAX,
				0, 10000);
		INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM.addChangeListener(rangeSliderListener);

		c.gridy++;
		visToggleSpots = new JButton("Toggle visibility of SPOTS");
		visToggleSpots.addActionListener(toggleSpotsVisibility);
		visToggleVols = new JButton("Toggle visibility of VOLUME");
		visToggleVols.addActionListener(toggleVolumeVisibility);
		//
		JPanel twoCenteredButtonsPlaceHolder = new JPanel();
		contentPane.add(twoCenteredButtonsPlaceHolder, c);
		//
		twoCenteredButtonsPlaceHolder.setLayout( new GridBagLayout() );
		GridBagConstraints bc = new GridBagConstraints();
		bc.fill = GridBagConstraints.HORIZONTAL;
		bc.weightx = 0.4;
		bc.gridx = 0;
		bc.insets = new Insets(0,0,0,20);
		twoCenteredButtonsPlaceHolder.add(visToggleSpots, bc);
		bc.gridx = 1;
		bc.insets = new Insets(0,20,0,0);
		twoCenteredButtonsPlaceHolder.add(visToggleVols, bc);

		// -------------- separator --------------
		c.gridy++;
		insertSeparator(c);

		c.gridy++;
		insertNote("Parameters of the spots in-painting into the Volume", c);

		c.gridy++;
		INTENSITY_OF_COLORS_APPLY = new JCheckBox("Enable spots in-painting into the Volume (visible only on next repainting of Volume)");
		insertCheckBox(INTENSITY_OF_COLORS_APPLY, c);

		c.gridy++;
		c.gridx = 0;
		insertLabel("Next repainting of Volume will be triggered:", c);
		//
		c.gridx = 1;
		UPDATE_VOLUME_AUTOMATICALLY = new JComboBox<>(new String[] {updVolMsgA, updVolMsgM});
		UPDATE_VOLUME_AUTOMATICALLY.addActionListener( updVolAutoListener );
		c.anchor = GridBagConstraints.LINE_END;
		insertRColumnItem(UPDATE_VOLUME_AUTOMATICALLY, c);
		c.anchor = GridBagConstraints.LINE_START;

		c.gridy++;
		c.gridx = 0;
		insertLabel("When repainting, draw colors at this nominal intensity:", c);
		//
		c.gridx = 1;
		INTENSITY_OF_COLORS = new SpinnerNumberModel(2400.0, 0.0, 65535.0, 50.0);
		insertSpinner(INTENSITY_OF_COLORS, (f) -> controlledBridge.INTENSITY_OF_COLORS = f, c);

		c.gridy++;
		c.gridx = 0;
		insertLabel("When repainting, multiply spots radii with:", c);
		//
		c.gridx = 1;
		SPOT_RADIUS_SCALE = new SpinnerNumberModel(3.0, 0.0, 50.0, 1.0);
		insertSpinner(SPOT_RADIUS_SCALE, (f) -> controlledBridge.SPOT_RADIUS_SCALE = f, c);

		c.gridy++;
		INTENSITY_OF_COLORS_BOOST = new JCheckBox("Enable enhancing of spot colors when repainting them into the Volume");
		insertCheckBox(INTENSITY_OF_COLORS_BOOST, c);

		c.gridy++;
		UPDATE_VOLUME_VERBOSE_REPORTS = new JCheckBox("Verbose/debug reporting during Volume repainting");
		insertCheckBox(UPDATE_VOLUME_VERBOSE_REPORTS, c);

		// -------------- button row --------------
		c.gridy++;
		c.gridx = 0;
		c.gridwidth = 1;
		c.insets = new Insets(10,sideSpace,sideSpace,sideSpace);
		JButton redrawBtn = new JButton("  Repaint now (the recently painted timepoint)  ");
		redrawBtn.addActionListener( (e) -> controlledBridge.updateSciviewColoringNow() );
		contentPane.add(redrawBtn, c);

		c.gridx = 1;
		c.anchor = GridBagConstraints.LINE_END;
		JButton closeBtn = new JButton("Close");
		closeBtn.addActionListener( (e) -> controlledBridge.detachControllingUI() );
		insertRColumnItem(closeBtn, c);
	}

	final int sideSpace = 15;
	final Insets noteSpace = new Insets(2,sideSpace,8,2*sideSpace);
	void insertNote(final String noteText, final GridBagConstraints c) {
		final int prevGridW = c.gridwidth;
		final Insets prevInsets = c.insets;

		c.gridwidth = 2;
		c.insets = noteSpace;
		c.weightx = 0.1;

		c.gridx = 0;
		contentPane.add( new JLabel(noteText), c);

		c.gridwidth = prevGridW;
		c.insets = prevInsets;
	}

	void insertLabel(final String labelText, final GridBagConstraints c) {
		c.weightx = 0.5;
		contentPane.add(new JLabel(labelText), c);
	}

	final Dimension spinnerMinDim = new Dimension(200,20);
	final Dimension spinnerMaxDim = new Dimension(1000,20);
	void insertSpinner(final SpinnerModel model,
	                   final Consumer<Float> updaterOnEvents,
	                   final GridBagConstraints c) {
		c.anchor = GridBagConstraints.LINE_END;
		insertRColumnItem(new JSpinner(model), c);
		c.anchor = GridBagConstraints.LINE_START;

		OwnerAwareSpinnerChangeListener l = new OwnerAwareSpinnerChangeListener(updaterOnEvents, model);
		model.addChangeListener(l);
		spinnerModelsWithListeners.add(l);
	}
	void insertRColumnItem(final JComponent item, final GridBagConstraints c) {
		item.setMinimumSize(spinnerMinDim);
		item.setPreferredSize(spinnerMinDim);
		item.setMaximumSize(spinnerMaxDim);
		c.weightx = 0.3;
		contentPane.add(item, c);
	}

	void insertCheckBox(final JCheckBox cbox, final GridBagConstraints c) {
		final int prevFill = c.fill;
		final int prevGridW = c.gridwidth;

		cbox.addItemListener( checkboxChangeListener );
		checkBoxesWithListeners.add(cbox);

		c.fill = GridBagConstraints.HORIZONTAL;
		c.gridwidth = 2;

		c.gridx = 0;
		c.weightx = 0.1;
		contentPane.add(cbox, c);

		c.fill = prevFill;
		c.gridwidth = prevGridW;
	}

	final Insets sepSpace = new Insets(8,sideSpace,8,sideSpace);
	void insertSeparator(final GridBagConstraints c) {
		final int prevFill = c.fill;
		final int prevGridW = c.gridwidth;
		final Insets prevInsets = c.insets;

		c.fill = GridBagConstraints.HORIZONTAL;
		c.gridwidth = 2;
		c.weightx = 0.1;
		c.insets = sepSpace;

		c.gridx = 0;
		contentPane.add( new JSeparator(JSeparator.HORIZONTAL), c);

		c.fill = prevFill;
		c.gridwidth = prevGridW;
		c.insets = prevInsets;
	}

	class OwnerAwareSpinnerChangeListener implements ChangeListener {
		final Consumer<Float> pushChangeToHere;
		final SpinnerModel observedSource;
		//
		OwnerAwareSpinnerChangeListener(Consumer<Float> pushChangeToHere,
		                                SpinnerModel fromThisItem) {
			this.pushChangeToHere = pushChangeToHere;
			this.observedSource = fromThisItem;
		}
		//
		@Override
		public void stateChanged(ChangeEvent changeEvent) {
			SpinnerNumberModel s = (SpinnerNumberModel)changeEvent.getSource();
			pushChangeToHere.accept( s.getNumber().floatValue() );
			if (controlledBridge.UPDATE_VOLUME_AUTOMATICALLY) controlledBridge.updateSciviewColoringNow();
		}
	}
	ItemListener checkboxChangeListener = new ItemListener() {
		@Override
		public void itemStateChanged(ItemEvent itemEvent) {
			JCheckBox cb = (JCheckBox) itemEvent.getSource();
			if (cb == INTENSITY_OF_COLORS_APPLY) {
				controlledBridge.INTENSITY_OF_COLORS_APPLY = cb.isSelected();
				if (controlledBridge.UPDATE_VOLUME_AUTOMATICALLY) controlledBridge.updateSciviewColoringNow();
			} else if (cb == INTENSITY_OF_COLORS_BOOST) {
				controlledBridge.INTENSITY_OF_COLORS_BOOST = cb.isSelected();
				if (controlledBridge.UPDATE_VOLUME_AUTOMATICALLY) controlledBridge.updateSciviewColoringNow();
			} else if (cb == UPDATE_VOLUME_VERBOSE_REPORTS) {
				controlledBridge.UPDATE_VOLUME_VERBOSE_REPORTS = cb.isSelected();
			}
		}
	};
	final java.util.List<OwnerAwareSpinnerChangeListener> spinnerModelsWithListeners = new ArrayList<>(10);
	final java.util.List<JCheckBox> checkBoxesWithListeners = new ArrayList<>(10);
	//
	//below is also defined: rangeSliderListener
	//
	final ActionListener updVolAutoListener = new ActionListener() {
		@Override
		public void actionPerformed(ActionEvent actionEvent) {
			boolean prevBridgeState = controlledBridge.UPDATE_VOLUME_AUTOMATICALLY;
			controlledBridge.UPDATE_VOLUME_AUTOMATICALLY
					= UPDATE_VOLUME_AUTOMATICALLY.getSelectedIndex() == 0;
			if (controlledBridge.UPDATE_VOLUME_AUTOMATICALLY && !prevBridgeState) {
				//NB: a protection "if" here because this listener can be triggered even
				//when one re-chooses (re-clicks) the already chosen element
				controlledBridge.updateSciviewColoringNow();
			}
		}
	};
	final ActionListener toggleSpotsVisibility = new ActionListener() {
		@Override
		public void actionPerformed(ActionEvent actionEvent) {
			boolean newState = !controlledBridge.sphereParent.getVisible();
			controlledBridge.setVisibilityOfSpots(newState);
		}
	};
	final ActionListener toggleVolumeVisibility = new ActionListener() {
		@Override
		public void actionPerformed(ActionEvent actionEvent) {
			boolean newState = !controlledBridge.redVolChannelNode.getVisible();
			controlledBridge.setVisibilityOfVolume(newState);
		}
	};

	/**
	 * Disable all listeners to make sure that, even if this UI window would ever
	 * be re-displayed, its controls could not control anything (and would throw
	 * NPEs if the controls would actually be used).
	 */
	public void deactivateAndForget() {
		//listeners tear-down here
		spinnerModelsWithListeners.forEach(c -> c.observedSource.removeChangeListener(c));
		checkBoxesWithListeners.forEach(c -> c.removeItemListener( checkboxChangeListener ));
		INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM.removeChangeListener(rangeSliderListener);
		UPDATE_VOLUME_AUTOMATICALLY.removeActionListener( updVolAutoListener );
		visToggleSpots.removeActionListener(toggleSpotsVisibility);
		visToggleVols.removeActionListener(toggleVolumeVisibility);
		this.controlledBridge = null;
	}

	public void updatePaneValues() {
		final boolean updVolAutoBackup = controlledBridge.UPDATE_VOLUME_AUTOMATICALLY;
		//temporarily disable because setting the controls trigger their listeners
		//that trigger (not all of them) the expensive volume updating
		controlledBridge.UPDATE_VOLUME_AUTOMATICALLY = false;

		INTENSITY_CONTRAST.setValue( controlledBridge.INTENSITY_CONTRAST );
		INTENSITY_SHIFT.setValue( controlledBridge.INTENSITY_SHIFT );
		INTENSITY_CLAMP_AT_TOP.setValue( controlledBridge.INTENSITY_CLAMP_AT_TOP );
		INTENSITY_GAMMA.setValue( controlledBridge.INTENSITY_GAMMA );

		INTENSITY_OF_COLORS.setValue( controlledBridge.INTENSITY_OF_COLORS );

		INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM
				.getSlider()
				.setValue( (int)controlledBridge.INTENSITY_RANGE_MIN );
		INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM
				.getRangeSlider()
				.setUpperValue( (int)controlledBridge.INTENSITY_RANGE_MAX );

		INTENSITY_OF_COLORS_APPLY.setSelected( controlledBridge.INTENSITY_OF_COLORS_APPLY );
		INTENSITY_OF_COLORS_BOOST.setSelected( controlledBridge.INTENSITY_OF_COLORS_BOOST );
		SPOT_RADIUS_SCALE.setValue( controlledBridge.SPOT_RADIUS_SCALE );

		UPDATE_VOLUME_AUTOMATICALLY.setSelectedItem(updVolAutoBackup ? updVolMsgA : updVolMsgM);
		UPDATE_VOLUME_VERBOSE_REPORTS.setSelected( controlledBridge.UPDATE_VOLUME_VERBOSE_REPORTS );

		controlledBridge.UPDATE_VOLUME_AUTOMATICALLY = updVolAutoBackup;
	}

	//int SOURCE_ID = 0;
	//int SOURCE_USED_RES_LEVEL = 0;

	SpinnerModel INTENSITY_CONTRAST;
	SpinnerModel INTENSITY_SHIFT;
	SpinnerModel INTENSITY_CLAMP_AT_TOP;
	SpinnerModel INTENSITY_GAMMA;

	SpinnerModel INTENSITY_OF_COLORS;

	AdjustableBoundsRangeSlider INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM;
	final ChangeListener rangeSliderListener = (l) -> {
		controlledBridge.INTENSITY_RANGE_MIN = INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM.getValue();
		controlledBridge.INTENSITY_RANGE_MAX = INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM.getUpperValue();
		controlledBridge.redVolChannelNode.setMinDisplayRange( controlledBridge.INTENSITY_RANGE_MIN );
		controlledBridge.redVolChannelNode.setMaxDisplayRange( controlledBridge.INTENSITY_RANGE_MAX );
	};
	//
	JButton visToggleSpots, visToggleVols;

	JCheckBox INTENSITY_OF_COLORS_APPLY;
	JCheckBox INTENSITY_OF_COLORS_BOOST;
	SpinnerModel SPOT_RADIUS_SCALE;

	static final String updVolMsgA = "Automatically";
	static final String updVolMsgM = "Manually";
	private JComboBox<String> UPDATE_VOLUME_AUTOMATICALLY;
	private JCheckBox UPDATE_VOLUME_VERBOSE_REPORTS;
}

File: src/main/kotlin/SciviewBridgeUI.kt
package org.mastodon.mamut

import graphics.scenery.utils.lazyLogger
import util.AdjustableBoundsRangeSlider
import util.GroupLocksHandling
import util.SphereLinkNodes
import java.awt.*
import java.awt.event.ActionListener
import java.util.function.Consumer
import javax.swing.*
import javax.swing.border.EmptyBorder
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

class SciviewBridgeUI(controlledBridge: SciviewBridge, populateThisContainer: Container) {
    var controlledBridge: SciviewBridge?
    val controlsWindowPanel: Container
    private val logger by lazyLogger()

    //int SOURCE_ID = 0;
    //int SOURCE_USED_RES_LEVEL = 0;
    lateinit var INTENSITY_CONTRAST: SpinnerModel
    lateinit var INTENSITY_SHIFT: SpinnerModel
    lateinit var INTENSITY_CLAMP_AT_TOP: SpinnerModel
    lateinit var INTENSITY_GAMMA: SpinnerModel
    lateinit var INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM: AdjustableBoundsRangeSlider
    //
    lateinit var visToggleSpots: JButton
    lateinit var visToggleVols: JButton
    lateinit var visToggleTracks: JButton
    lateinit var linkRangeBackwards: SpinnerModel
    lateinit var linkRangeForwards: SpinnerModel
    lateinit var autoIntensityBtn: JToggleButton
    lateinit var lockGroupHandler: GroupLocksHandling
    lateinit var linkColorSelector: JComboBox<String>
    lateinit var volumeColorSelector: JComboBox<String>

    // -------------------------------------------------------------------------------------------
    private fun populatePane() {
        val bridge = this.controlledBridge ?: throw IllegalStateException("The passed bridge cannot be null.")

        controlsWindowPanel.setLayout(GridBagLayout())
        val c = GridBagConstraints()
        c.anchor = GridBagConstraints.LINE_START
        c.fill = GridBagConstraints.HORIZONTAL
        val mastodonRowPlaceHolder = JPanel()
        mastodonRowPlaceHolder.setLayout(GridBagLayout())
        val mc = GridBagConstraints()
        mc.anchor = GridBagConstraints.LINE_START
        mc.fill = GridBagConstraints.HORIZONTAL
        mc.insets = Insets(0, 0, 0, 0)
        mc.weightx = 0.2
        mc.gridx = 0
        lockGroupHandler = GroupLocksHandling(bridge, bridge.mastodon)
        mastodonRowPlaceHolder.add(lockGroupHandler.createAndActivate()!!, mc)
        mc.weightx = 0.6
        mc.gridx = 1
        val openBdvBtn = JButton("Open synced Mastodon BDV")
        openBdvBtn.addActionListener { bridge.openSyncedBDV() }
        mastodonRowPlaceHolder.add(openBdvBtn, mc)
        //
        c.gridy = 0
        c.gridwidth = 2
        c.gridx = 0
        c.weightx = 0.1
        c.insets = Insets(4, sideSpace, 8, sideSpace - 2)
        controlsWindowPanel.add(mastodonRowPlaceHolder, c)
        c.gridy++
        c.insets = Insets(2, sideSpace, 2, sideSpace)
        controlsWindowPanel.add(
            JLabel("Volume pixel values 'v' are processed linearly, normalized, gamma, scaled back:"),
            c
        )
        c.gridwidth = 1
        c.gridy++
        insertNote("   pow( min(contrast*v +shift, not_above)/not_above, gamma ) *not_above", c)
        c.gridy++
        c.gridx = 0
        insertLabel("Apply on Volume this contrast factor:", c)
        //
        c.gridx = 1
        INTENSITY_CONTRAST = SpinnerNumberModel(1.0, -100.0, 100.0, 0.5)
        insertSpinner(INTENSITY_CONTRAST, { f: Float -> bridge.intensity.contrast = f }, c)
        c.gridy++
        c.gridx = 0
        insertLabel("Apply on Volume this shifting bias:", c)
        //
        c.gridx = 1
        INTENSITY_SHIFT = SpinnerNumberModel(0.0, -65535.0, 65535.0, 50.0)
        insertSpinner(INTENSITY_SHIFT, { f: Float -> bridge.intensity.shift = f }, c)
        c.gridy++
        c.gridx = 0
        insertLabel("Apply on Volume this gamma level:", c)
        //
        c.gridx = 1
        INTENSITY_GAMMA = SpinnerNumberModel(1.0, 0.1, 3.0, 0.1)
        insertSpinner(INTENSITY_GAMMA, { f: Float -> bridge.intensity.gamma = f }, c)
        c.gridy++
        c.gridx = 0
        insertLabel("Clamp all voxels so that their values are not above:", c)
        //
        c.gridx = 1
        INTENSITY_CLAMP_AT_TOP = SpinnerNumberModel(700.0, 0.0, 65535.0, 50.0)
        insertSpinner(
            INTENSITY_CLAMP_AT_TOP,
            { f: Float -> bridge.intensity.clampTop = f },
            c
        )

        // -------------- separator --------------
        c.gridy++
        insertSeparator(c)
        c.gridy++
        insertNote("Shortcuts to the standard sciview view controls, incl. Volume intensity mapping", c)
        c.gridy++
        val sliderBarPlaceHolder = JPanel()
        c.gridx = 0
        c.gridwidth = 2
        controlsWindowPanel.add(sliderBarPlaceHolder, c)
        //
        INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM = AdjustableBoundsRangeSlider.createAndPlaceHere(
            sliderBarPlaceHolder,
            bridge.intensity.rangeMin.toInt(),
            bridge.intensity.rangeMax.toInt(),
            0,
            10000
        )
        INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM.addChangeListener(rangeSliderListener)


        // links window range
        c.gridy++
        c.gridwidth = 2
        c.gridx = 0
        insertLabel("Link window range backwards", c)
        c.gridx = 1
        linkRangeBackwards = SpinnerNumberModel(bridge.mastodon.maxTimepoint, 0, bridge.mastodon.maxTimepoint, 1)
        insertSpinner(
            linkRangeBackwards,
            {
                f: Float -> bridge.sphereLinkNodes.linkBackwardRange = f.toInt()
                bridge.sphereLinkNodes.updateLinkVisibility(bridge.lastTpWhenVolumeWasUpdated)
            },
            c)

        c.gridy++
        c.gridx = 0
        insertLabel("Link window range forwards:", c)
        c.gridx = 1
        linkRangeForwards = SpinnerNumberModel(bridge.mastodon.maxTimepoint, 0, bridge.mastodon.maxTimepoint, 1)
        insertSpinner(
            linkRangeForwards,
            {
                f: Float -> bridge.sphereLinkNodes.linkForwardRange = f.toInt()
                bridge.sphereLinkNodes.updateLinkVisibility(bridge.lastTpWhenVolumeWasUpdated)
            },
            c
        )


        // color parameters
        c.gridy++
        c.gridwidth = 4
        val colorPlaceholder = JPanel()
        c.gridx = 0
        controlsWindowPanel.add(colorPlaceholder, c)
        colorPlaceholder.setLayout(GridBagLayout())
        val coloringRow = GridBagConstraints()
        coloringRow.fill = GridBagConstraints.HORIZONTAL
        coloringRow.gridx = 0
        colorPlaceholder.add(Label("Link colors: "), coloringRow)
        coloringRow.gridx = 1
        // add the first choice of the list manually
        val linkColorChoices = mutableListOf("By Spot")
        // get the rest of the LUTs from sciview and clean up their names
        val availableLUTs = bridge.sciviewWin.getAvailableLUTs() as MutableList<String>
        for (i in availableLUTs.indices) {
            availableLUTs[i] = availableLUTs[i].removeSuffix(".lut")
        }
        linkColorChoices.addAll(availableLUTs)

        // create dropdown menu for link LUTs
        linkColorSelector = JComboBox(linkColorChoices.toTypedArray())
        linkColorSelector.setSelectedItem("Fire")
        colorPlaceholder.add(linkColorSelector, coloringRow)
        linkColorSelector.addActionListener(chooseLinkColormap)

        // Volume colors
        coloringRow.gridx = 2
        coloringRow.insets = Insets(0, 10, 0, 0)
        colorPlaceholder.add(Label("Volume colors: "), coloringRow)
        coloringRow.gridx = 3
        volumeColorSelector = JComboBox(availableLUTs.toTypedArray())
        volumeColorSelector.setSelectedItem("Grays")
        colorPlaceholder.add(volumeColorSelector, coloringRow)
        volumeColorSelector.addActionListener(chooseVolumeColormap)

        // the four toggle buttons
        c.gridy++
        visToggleSpots = JButton("Toggle spots")
        visToggleSpots.addActionListener(toggleSpotsVisibility)
        visToggleVols = JButton("Toggle volume")
        visToggleVols.addActionListener(toggleVolumeVisibility)
        visToggleTracks = JButton("Toggle tracks")
        visToggleTracks.addActionListener(toggleTrackVisivility)

        autoIntensityBtn = JToggleButton("Auto Intensity", bridge.isVolumeAutoAdjust)
        autoIntensityBtn.addActionListener(autoAdjustIntensity)
        //
        val fourButtonsPlaceholder = JPanel()
        controlsWindowPanel.add(fourButtonsPlaceholder, c)
        //
        fourButtonsPlaceholder.setLayout(GridBagLayout())
        val buttonRow = GridBagConstraints()
        buttonRow.fill = GridBagConstraints.HORIZONTAL
        buttonRow.anchor = GridBagConstraints.WEST
        buttonRow.weightx = 0.4
        buttonRow.gridx = 0
//        bc.insets = Insets(0, 20, 0, 0)
        fourButtonsPlaceholder.add(autoIntensityBtn, buttonRow)
        buttonRow.gridx = 1
        buttonRow.insets = Insets(0, 20, 0, 0)
        fourButtonsPlaceholder.add(visToggleSpots, buttonRow)
        buttonRow.gridx = 2
        buttonRow.insets = Insets(0, 20, 0, 0)
        fourButtonsPlaceholder.add(visToggleVols, buttonRow)
        buttonRow.gridx = 3
        buttonRow.insets = Insets(0, 20, 0, 0)
        fourButtonsPlaceholder.add(visToggleTracks, buttonRow)

        // -------------- close button row --------------
        c.gridy++
        c.gridx = 1
        c.gridwidth = 1
        c.anchor = GridBagConstraints.EAST
        val eyetrackingButton = JButton("Start Eye Tracking")
        eyetrackingButton.addActionListener { bridge.launchEyeTracking() }
        controlsWindowPanel.add(eyetrackingButton, c)
        c.insets = Insets(0, 20, 0, 0)
        c.gridx = 2
        val closeBtn = JButton("Close")
        closeBtn.addActionListener { bridge.detachControllingUI() }
        c.insets = Insets(0, 0, 0, 15)
        controlsWindowPanel.add(closeBtn, c)
    }

    val sideSpace = 15
    val noteSpace = Insets(2, sideSpace, 8, 2 * sideSpace)
    fun insertNote(noteText: String?, c: GridBagConstraints) {
        val prevGridW = c.gridwidth
        val prevInsets = c.insets
        c.gridwidth = 2
        c.insets = noteSpace
        c.weightx = 0.1
        c.gridx = 0
        controlsWindowPanel.add(JLabel(noteText), c)
        c.gridwidth = prevGridW
        c.insets = prevInsets
    }

    fun insertLabel(labelText: String?, c: GridBagConstraints) {
        c.weightx = 0.5
        controlsWindowPanel.add(JLabel(labelText), c)
    }

    val spinnerMinDim = Dimension(40, 20)
    val spinnerMaxDim = Dimension(200, 20)
    fun insertSpinner(
        model: SpinnerModel,
        updaterOnEvents: Consumer<Float>,
        c: GridBagConstraints
    ) {
        c.anchor = GridBagConstraints.LINE_END
        insertRColumnItem(JSpinner(model), c)
        c.anchor = GridBagConstraints.LINE_START
        val l = OwnerAwareSpinnerChangeListener(updaterOnEvents, model)
        model.addChangeListener(l)
        spinnerModelsWithListeners.add(l)
    }

    fun insertRColumnItem(item: JComponent, c: GridBagConstraints) {
        item.minimumSize = spinnerMinDim
        item.preferredSize = spinnerMinDim
        item.maximumSize = spinnerMaxDim
        c.weightx = 0.3
        controlsWindowPanel.add(item, c)
    }

    fun insertCheckBox(cbox: JCheckBox, c: GridBagConstraints) {
        val prevFill = c.fill
        val prevGridW = c.gridwidth
        checkBoxesWithListeners.add(cbox)
        c.fill = GridBagConstraints.HORIZONTAL
        c.gridwidth = 2
        c.gridx = 0
        c.weightx = 0.1
        controlsWindowPanel.add(cbox, c)
        c.fill = prevFill
        c.gridwidth = prevGridW
    }

    val sepSpace = Insets(8, sideSpace, 8, sideSpace)
    fun insertSeparator(c: GridBagConstraints) {
        val prevFill = c.fill
        val prevGridW = c.gridwidth
        val prevInsets = c.insets
        c.fill = GridBagConstraints.HORIZONTAL
        c.gridwidth = 2
        c.weightx = 0.1
        c.insets = sepSpace
        c.gridx = 0
        controlsWindowPanel.add(JSeparator(JSeparator.HORIZONTAL), c)
        c.fill = prevFill
        c.gridwidth = prevGridW
        c.insets = prevInsets
    }

    inner class OwnerAwareSpinnerChangeListener //
        (
        val pushChangeToHere: Consumer<Float>,
        val observedSource: SpinnerModel
    ) : ChangeListener {
        //
        override fun stateChanged(changeEvent: ChangeEvent) {
            val bridge = this@SciviewBridgeUI.controlledBridge ?: throw IllegalStateException("Bridge is null.")
            val s = changeEvent.source as SpinnerNumberModel
            pushChangeToHere.accept(s.number.toFloat())
            if (bridge.updateVolAutomatically) bridge.updateVolume()
        }
    }

    val spinnerModelsWithListeners: MutableList<OwnerAwareSpinnerChangeListener> = ArrayList(10)
    val checkBoxesWithListeners: MutableList<JCheckBox> = ArrayList(10)

    val chooseLinkColormap = ActionListener { _ ->
        when (linkColorSelector.selectedItem) {
            "By Spot" -> {
                controlledBridge.sphereLinkNodes.currentColorMode = SphereLinkNodes.ColorMode.SPOT
                logger.info("Coloring links by spot color")
            }
            else -> {
                controlledBridge.sphereLinkNodes.currentColorMode = SphereLinkNodes.ColorMode.LUT
                controlledBridge.sphereLinkNodes.setLUT("${linkColorSelector.selectedItem}.lut")
                logger.info("Coloring links with LUT ${linkColorSelector.selectedItem}")
            }
        }
        controlledBridge.sphereLinkNodes.updateLinkColors(controlledBridge.recentColorizer ?: controlledBridge.noTSColorizer)
    }

    val chooseVolumeColormap = ActionListener {
        controlledBridge.sciviewWin.setColormap(controlledBridge.volumeNode, "${volumeColorSelector.selectedItem}.lut")
        logger.info("Coloring volume with LUT ${volumeColorSelector.selectedItem}")
    }

    val toggleSpotsVisibility = ActionListener {
        val newState = !controlledBridge.sphereParent.visible
        controlledBridge.sphereParent.visible = newState
    }
    val toggleVolumeVisibility = ActionListener {
        val newState = !controlledBridge.volumeNode.visible
        controlledBridge.setVisibilityOfVolume(newState)
    }
    val toggleTrackVisivility = ActionListener {
        val newState = !controlledBridge.linkParent.visible
        controlledBridge.linkParent.visible = newState
    }

    val autoAdjustIntensity = ActionListener {
        controlledBridge.autoAdjustIntensity()
    }

    /**
     * Disable all listeners to make sure that, even if this UI window would ever
     * be re-displayed, its controls could not control anything (and would throw
     * NPEs if the controls would actually be used).
     */
    fun deactivateAndForget() {
        //listeners tear-down here
        lockGroupHandler.deactivate()
        spinnerModelsWithListeners.forEach { c: OwnerAwareSpinnerChangeListener ->
            c.observedSource.removeChangeListener(
                c
            )
        }
        INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM.removeChangeListener(rangeSliderListener)
        visToggleSpots.removeActionListener(toggleSpotsVisibility)
        visToggleVols.removeActionListener(toggleVolumeVisibility)
        autoIntensityBtn.removeActionListener(autoAdjustIntensity)
        controlledBridge = null
    }

    fun updatePaneValues() {
        val bridge = this.controlledBridge ?: throw IllegalStateException("Bridge is null.")
        val updVolAutoBackup = bridge.updateVolAutomatically
        //temporarily disable because setting the controls trigger their listeners
        //that trigger (not all of them) the expensive volume updating

        bridge.updateVolAutomatically = false
        INTENSITY_CONTRAST.value = bridge.intensity.contrast
        INTENSITY_SHIFT.value = bridge.intensity.shift
        INTENSITY_CLAMP_AT_TOP.value = bridge.intensity.clampTop
        INTENSITY_GAMMA.value = bridge.intensity.gamma
        val upperValBackup = bridge.intensity.rangeMax

        INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM
            .rangeSlider
            .value = bridge.intensity.rangeMin.toInt()
        //NB: this triggers a "value changed listener" which updates _both_ the value and upperValue,
        //    which resets the value with the new one (so no change in the end) but clears upperValue
        //    to the value the dialog was left with (forgets the new upperValue effectively)
        bridge.intensity.rangeMax = upperValBackup
        INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM
            .rangeSlider
            .upperValue = bridge.intensity.rangeMax.toInt()

        autoIntensityBtn.isSelected = bridge.isVolumeAutoAdjust
        bridge.updateVolAutomatically = updVolAutoBackup
    }


    val rangeSliderListener = ChangeListener {
        controlledBridge.intensity.rangeMin = INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM.value.toFloat()
        controlledBridge.intensity.rangeMax = INTENSITY_RANGE_MINMAX_CTRL_GUI_ELEM.upperValue.toFloat()
        controlledBridge.volumeNode.minDisplayRange = controlledBridge.intensity.rangeMin
        controlledBridge.volumeNode.maxDisplayRange = controlledBridge.intensity.rangeMax
    }



    init {
        this.controlledBridge = controlledBridge
        controlsWindowPanel = populateThisContainer
        populatePane()
        updatePaneValues()
    }

    companion object {
        const val updVolMsgA = "Automatically"
        const val updVolMsgM = "Manually"
    }
}